[Nicholas Sun] iP

NicsunXnus opened the pull request:
### Friendly Friend ChatBot
> "The Mother of all ChatBots (___not really___)" - ***Me***
> "Efficiency is doing things right. Effectiveness is doing the right thing. " - Zig Zigler ([Source](https://www.quotetab.com/quote/by-zig-ziglar/efficiency-is-doing-things-right-effectiveness-is-doing-the-right-thing))

Duke ChatBot frees your mind of having to remember things you need to do. It's,
- GUI-based
- easy to learn

All you need to do is:
1. download it from [here](https://github.com/NicsunXnus/ip). 
2. double click it.
3. add your tasks.
4. Let it manage your tasks for you :+1::+1:

Features:
- [x] Mark to-do tasks
- [x] Mark deadline tasks
- [x] Mark event tasks
- [x] Find and delete tasks
- [x] Save tasks into a file
- [ ] Completed 

If you are a Java programmer, you can use it to practise Java too. Here is the ```main``` method:
```ruby
public class Main {
    public static void main(String[] args) {
        new Duke("C:/repos/ip/data/tasks.txt").run();
    }
}
```

Commit:
Add Gradle support

NicsunXnus committed:
Created Action class with its methods(getAction,toString)

NicsunXnus committed:
Added isBye method

NicsunXnus committed:
Fnished level 1

NicsunXnus committed:
Added listOut method(which lists out the things inputted so far) in Duke.

NicsunXnus committed:
Finished level 2

NicsunXnus committed:
Added methods to Action.java

NicsunXnus committed:
Finished level 3

NicsunXnus committed:
Added Deadline,Todo classes

NicsunXnus committed:
Finished up level 4

NicsunXnus committed:
Finished with A-TextUiTesting

NicsunXnus committed:
Added/Updated DukeException class

NicsunXnus committed:
Finished with Level-5.
Added method testForUnknown.

NicsunXnus committed:
Minor changes

NicsunXnus committed:
Added isDelete method to Action class.
Added if clause for delete task in Duke class.
Finished level 6

NicsunXnus committed:
import java.io.File

NicsunXnus committed:
finish level 7

NicsunXnus committed:
finish level 8

NicsunXnus committed:
Merge branch 'branch-Level-7' into branch-Level-8

NicsunXnus committed:
Merge remote-tracking branch 'nus-cs2103-AY2122S2/add-gradle-support' into branch-Level-8

NicsunXnus committed:
revert merge to remote repo

NicsunXnus committed:
Revert "Merge branch 'branch-Level-7' into branch-Level-8"

This reverts commit f2f3bf6b244edbaec7fd52268388ba6924986c3f, reversing
changes made to a4a95dd6a653997ecd4ecf9e985ee073903b7e80.

revert merge to level-branch-7

NicsunXnus committed:
Merge branch 'branch-Level-7'

NicsunXnus committed:
Merge branch 'branch-Level-8'

NicsunXnus committed:
Halfway through OOP

NicsunXnus committed:
old duke file

NicsunXnus committed:
finished more "OOP"

NicsunXnus committed:
finished "A=Packages"

NicsunXnus committed:
finished "A-Junit"

NicsunXnus committed:
finished javadoc additions

NicsunXnus committed:
finish coding standard changes

NicsunXnus committed:
finish find

NicsunXnus committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/duke/Parser.java
#	src/main/java/duke/Storage.java
#	src/main/java/duke/Ui.java
#	src/main/java/duke/command/Command.java
#	src/main/java/duke/command/ExitCommand.java
#	src/main/java/duke/command/ListCommand.java
#	src/main/java/duke/command/MarkCommand.java
#	src/main/java/duke/command/UnmarkCommand.java

NicsunXnus committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/Parser.java
#	src/main/java/duke/TaskList.java
#	src/main/java/duke/action/Action.java

NicsunXnus committed:
brush up after merges

NicsunXnus committed:
create the gradle files

NicsunXnus committed:
commit A-Checkstyle

NicsunXnus committed:
edit packages for files

NicsunXnus committed:
gui

NicsunXnus committed:
improve gui

NicsunXnus committed:
adjust LocatDate to localdatetime

NicsunXnus committed:
add new junit test

NicsunXnus committed:
add comments to classes

dandaandaaaaaan commented on src/test/ActionTest.java:
```diff
@@ -0,0 +1,29 @@
+    }
+
+    @Test
+    public void testSetDone() {
```
```suggestion
    public void getStatus_markAction_xReturned() {
```
test function name could be improved into the 3 part `featureUnderTest_testScenario_expectedBehaviour`

dandaandaaaaaan commented on src/test/ActionTest.java:
```diff
@@ -0,0 +1,29 @@
+    }
+
+    @Test
+    public void testSetUndone() {
```
```suggestion
    public void getStatus_markAction_spaceReturned() {
```

NicsunXnus replied:
Alright noted!

dandaandaaaaaan commented on src/test/DeadlineAndEventDateTimeParsingTest.java:
```diff
@@ -0,0 +1,28 @@
+
+public class DeadlineAndEventDateTimeParsingTest {
+    @Test
+    public void testDeadlineConstructor() {
```
```suggestion
    public void createDeadline_validTask_success() {
```

dandaandaaaaaan commented on src/test/DeadlineAndEventDateTimeParsingTest.java:
```diff
@@ -0,0 +1,28 @@
+    }
+
+    @Test
+    public void testEventConstructor() {
```
```suggestion
    public void createEvent_validTask_success() {
```

dandaandaaaaaan commented on src/main/java/duke/action/Action.java:
```diff
@@ -0,0 +1,82 @@
+     * @param string task
+     * @param status done status of task
+     */
+    public Action(String string, boolean status)  {
```
```suggestion
    public Action(String string, boolean status) {
```
Coding standard spacing

dandaandaaaaaan commented on src/main/java/duke/action/Deadline.java:
```diff
@@ -0,0 +1,76 @@
+     * @param by date and time values
+     * @param bool done status of task with deadline
+     */
+    public Deadline(String task, LocalDateTime by, boolean bool)  {
```
```suggestion
    public Deadline(String task, LocalDateTime by, boolean bool) {
```
Coding standard spacing

dandaandaaaaaan commented on src/main/java/duke/action/Deadline.java:
```diff
@@ -0,0 +1,76 @@
+     * @return marked Deadline object
+     */
+    @Override
+    public Action setDone()  {
```
```suggestion
    public Action setDone() {
```
Coding standard spacing

dandaandaaaaaan commented on src/main/java/duke/action/Deadline.java:
```diff
@@ -0,0 +1,76 @@
+     * @return unmarked Deadline object
+     */
+    @Override
+    public Action setUnDone()  {
```
```suggestion
    public Action setUnDone() {
```
Coding standard spacing

dandaandaaaaaan commented on src/main/java/duke/action/Event.java:
```diff
@@ -0,0 +1,74 @@
+     * @param date date and time values of the event
+     * @param bool done status of the event
+     */
+    public Event(String task, LocalDateTime date, boolean bool){
```
```suggestion
    public Event(String task, LocalDateTime date, boolean bool) {
```
Coding standard spacing

dandaandaaaaaan commented on src/main/java/duke/action/Event.java:
```diff
@@ -0,0 +1,74 @@
+     * @return marked Event object
+     */
+    @Override
+    public Action setDone()  {
```
```suggestion
    public Action setDone() {
```
Coding standard spacing

dandaandaaaaaan commented on src/main/java/duke/action/Event.java:
```diff
@@ -0,0 +1,74 @@
+     * @return unmarked Event object
+     */
+    @Override
+    public Action setUnDone()  {
```
```suggestion
    public Action setUnDone() {
```
Coding standard spacing

dandaandaaaaaan commented on src/main/java/duke/action/Todo.java:
```diff
@@ -0,0 +1,54 @@
+     * Constructs a new Todo object with a String variable s.
+     * @param s task todo
+     */
+    public Todo(String s)  {
```
```suggestion
    public Todo(String s) {
```
Coding standard spacing

dandaandaaaaaan commented on src/main/java/duke/action/Todo.java:
```diff
@@ -0,0 +1,54 @@
+     * @param s task todo
+     * @param bool done status of task todo
+     */
+    public Todo(String s, boolean bool)  {
```
```suggestion
    public Todo(String s, boolean bool) {
```
Coding standard spacing

dandaandaaaaaan commented on src/main/java/duke/action/Todo.java:
```diff
@@ -0,0 +1,54 @@
+     * @return marked Todo object
+     */
+    @Override
+    public Action setDone()  {
```
```suggestion
    public Action setDone() {
```
Coding standard spacing

dandaandaaaaaan reviewed (commented):
Most of the naming schemes here are alright, highlighted a few but just mainly tests. Also highlighted some coding standard violations with the spacings, but overall apart from minor nits LGTM!

Cyolune commented on src/main/java/duke/command/AddCommand.java:
```diff
@@ -0,0 +1,73 @@
+     * @param details details of the task etc; anything after the command
+     */
+    public AddCommand(String command, String details) {
+        if (command.equals("todo")) {
+            action = new Todo(details);
+        } else {
+            if (command.equals("deadline")) {
+                String[] thisAction = details.split("/by ");
+                String act = thisAction[0];
+                String date = thisAction[1];
+                action = new Deadline(act, date);
+            } else {
+                String[] thisAction = details.split("/at ");
+                String act = thisAction[0];
+                String date = thisAction[1];
+                action = new Event(act, date);
+            }
```
Is there a reason for the if else checks to be on different indentation levels? Maybe having them on the same levels would make the code more maintainable.

Cyolune commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,26 @@
+     * @param ui user interface
+     * @param storage file storage
+     */
+    public abstract String execute(TaskList taskList, Ui ui, Storage storage);
```
Would having the method name be `executeCommand` make it clearer to tell what its doing?

Cyolune commented on src/main/java/duke/ui/TaskList.java:
```diff
@@ -0,0 +1,116 @@
+     * according to the given index from the input
+     * @param taskNo index of task
+     */
+    public void setUnDone(int taskNo) {
```
Since undone is a single word, should the method name be `setUndone`?

Cyolune commented on src/main/java/duke/ui/Parser.java:
```diff
@@ -0,0 +1,65 @@
+
+        String[] twoWords = command.split(" ", 2);
+        String firstWord = twoWords[0];
+        if (firstWord.equals("delete") || firstWord.contains("mark")) {
+            int number = Integer.parseInt(twoWords[1]);
+            if (firstWord.equals("delete")) {
+                return new DeleteCommand(number);
+            }
+
+            if (firstWord.equals("mark")) {
+                return new MarkCommand(number);
+            }
+
+            if (firstWord.equals("unmark")) {
+                return new UnmarkCommand(number);
+            }
+        }
+
+        if (firstWord.equals("find")) {
+            String task = twoWords[1];
+            return new FindCommand(task);
+        }
+
+        if (!firstWord.equals("deadline") && !firstWord.equals("todo") && !firstWord.equals("event")) {
+            throw new DukeException("Unknown Command");
+        }
+
+        String details = twoWords[1];
+        return new AddCommand(firstWord, details);
```
Would it be clearer for this entire block to be if elses that parses each command, then have the last else throw the `DukeException`?

Cyolune reviewed (commented):
LGTM

NicsunXnus committed:
add try-catch blocks in parser class to catch datetimeparse exception

NicsunXnus committed:
change build.gradle

NicsunXnus committed:
GUI window: Add name and a icon

GUI window looks lacking design wise.

A logo and a name would make the GUI window realistic as well as aesthetic.

Let's,
1) update the start() method in the Main class
2) use javafx setTitle method to set the title to "Duke"
3) use javafx getIcons method to set the image to "/Images/red.png"

NicsunXnus committed:
Duke: remove text-ui

There is now a GUI in place of the text ui.

Let's,
1) remove the run method in Duke
2) remove the main method in Duke

NicsunXnus committed:
add assert lines to MarkCommand and UnmarkCommand

NicsunXnus committed:
remove duke.txt

NicsunXnus committed:
remove unused codes from Ui

NicsunXnus committed:
Parser: if-else statements are shortened

The parse method is too long and some if-else statements could be packaged into methods and switch statements.

NicsunXnus committed:
Merge pull request #1 from NicsunXnus/branch-A-Assertions

Branch a assertions

NicsunXnus committed:
Merge pull request #2 from NicsunXnus/branch-A-CodeQuality

Branch a code quality

NicsunXnus committed:
add help button

NicsunXnus committed:
add help command

Changes:
1) Parser: Add new conditional statement to parse for "help" input
2) MainWindow: Add method provideAboutFunctionality() to provide a link to the help webpage
3) HelpCommand: Add new command class to be responsible for executing "help" input

NicsunXnus committed:
make code adhere more to coding standard

juayhee commented on src/main/java/duke/gui/MainWindow.java:
```diff
@@ -87,10 +82,9 @@ private void handleUserInput() throws URISyntaxException, IOException {
      */
     @FXML
-    private void handleAboutAction(final ActionEvent event) throws URISyntaxException, IOException {
+    private void handleAboutAction() throws URISyntaxException, IOException {
```
I really like the use of abstraction, even when handling exceptions you choose to pass it off to another function, which is quite clean.

juayhee commented on src/main/java/duke/ui/Parser.java:
```diff
@@ -28,44 +28,39 @@
     }
 
-    public static void checkForEmptyCommands(String command) throws DukeException {
+    /**
+     * Checks for inputs that are either empty or are missing details.
+     * Incomplete commands only applies to todo/deadline/event commands.
+     * @param command user input
+     * @throws DukeException throws "Empty Command" or "XXX cannot be empty"
+     */
```
Love the very informative javadocs

juayhee commented on src/main/java/Deadline.java:
```diff
@@ -1,28 +1,37 @@
     }
 
-    public Deadline(String task, String by, boolean bool)  {
+    public Deadline(String task, LocalDate by, boolean bool)  {
```
Perhaps the boolean variable can be renamed to something more descriptive?

NicsunXnus committed:
improve the GUI

NicsunXnus committed:
change name of GUI

NicsunXnus committed:
improve coding standard

TaskList: Added some exceptions handling within setDone, setUnDone, delete

NicsunXnus committed:
add GUI screenshot

NicsunXnus committed:
Add files via upload

NicsunXnus committed:
test

NicsunXnus committed:
Delete Ui.png.jpg

NicsunXnus committed:
Update README.md

NicsunXnus committed:
Set theme jekyll-theme-time-machine

NicsunXnus committed:
Minor changes to Deadline and Event

Changed dateTimeFormat in toString() from "MMM d yyyy, H:m" to "MMM d yyyy, HH:mm"

NicsunXnus committed:
Delete Ui.png.jpg

NicsunXnus committed:
reupload Ui.png

NicsunXnus committed:
Update README.md

NicsunXnus committed:
Update README.md

NicsunXnus committed:
Update README.md

NicsunXnus committed:
Added new classes to read saved data from last session.

Classes added: SavedDataParser, ReadFile

Also, made some coding standard changes.

NicsunXnus committed:
Merge branch 'master' of https://github.com/NicsunXnus/ip

NicsunXnus committed:
Added new screen message to GUI

NicsunXnus committed:
Main: Changes to file path
Storage:Changes to file path

NicsunXnus committed:
Changed the file path to be more OS independent

Removed test.java due to redundancy.

NicsunXnus committed:
Add files via upload

NicsunXnus committed:
Added stylesheet.css to src/main/resources/view

NicsunXnus committed:
Delete Ui.png

NicsunXnus committed:
Add files via upload

NicsunXnus committed:
Delete Ui.png

NicsunXnus committed:
Add files via upload

NicsunXnus committed:
reupload Ui.png

NicsunXnus committed:
Merge remote-tracking branch 'origin/master'

NicsunXnus committed:
MainwWindow.fxml:Removed prefHeight from VBox

Rationale: Having prefHeight for VBox made it impossible for the all the tasks in the dialogBox to be displayed.

Changes: Removed prefHeight value and replace it with a minHeight with same value

NicsunXnus committed:
MainwWindow: Changed the startup messages to look better on startup of GUI.

NicsunXnus committed:
ClearCommand: Added new command extension for dealing with clearing of tasks from tasklist

Rationale: Having a command to clear tasks makes the GUI more useful.

Changes: Added required code to Parser, and clear() method to TaskList

NicsunXnus committed:
build.gradle: Changed version to 2.1

NicsunXnus committed:
Minor commit: Java coding standard things

NicsunXnus committed:
Update README.md

NicsunXnus committed:
Update README.md

NicsunXnus committed:
Storage: Changed save method to deal with empty tasklist

Changes: ClearCommand, DeleteCommand,ExitCommand, Storage

NicsunXnus committed:
Merge branch 'master' of https://github.com/NicsunXnus/ip

NicsunXnus committed:
SavedDataParser: Add a new method isMarkedTask()

Rationale:
There was a bug where task marked as done was not saved properly in the text file (thanks Charisma). This was due to the parse() only adding unmarked tasks back to the tasklist despite their marked status. parse() is then modified accordingly to include isMarkedTask() checks. Marked tasks are added accordingly.

NicsunXnus committed:
Parser: Added 2 methods to deal with incorrect inputs

Rationale:
Incorrect Deadline and Event inputs were not detected by the GUI and were not displayed. This might lead to confusion amongst users.

Changes:
1) checkForCorrectDeadlineInput() and checkForCorrectEventInput() were created to throw an exception if an incorrect input is found.
2) The exception is then displayed on the GUI to guide the user.

NicsunXnus committed:
Update README.md
